File: packages/typediagram/src/tdbin/writer.ts
import { ok, type Result } from "../result.js";
import { tdbinErr } from "./error.js";
import {
  ELEM_BIT,
  ELEM_BYTE,
  ELEM_COMPOSITE,
  ELEM_EIGHT_BYTES,
  ELEM_POINTER,
  encodeList,
  encodeStruct,
  relOffset,
} from "./pointer.js";
import type { StructCodec, TdbinError, Writer } from "./types.js";
import { WORD_BITS, WORD_BYTES, utf8Encode, wordsToBytes } from "./word.js";

const MAX_WORDS = 1 << 26;

export const createWriter = (): Writer => ({ body: [] });

export const message = <T>(codec: StructCodec<T>, value: T): Result<Uint8Array, TdbinError> => {
  const writer: Writer = { body: [0n] };
  const root = 0;
  const body = reserve(writer, codec.dataWords + codec.ptrWords);
  if (!body.ok) {
    return body;
  }
  const written = codec.write(writer, body.value, value);
  if (!written.ok) {
    return written;
  }
  const ptr = structPtr(body.value, root, codec.dataWords, codec.ptrWords);
  if (!ptr.ok) {
    return ptr;
  }
  writer.body[root] = ptr.value;
  return ok(wordsToBytes(writer.body));
};

export const scalar = (writer: Writer, at: number, slot: number, bits: bigint): Result<void, TdbinError> =>
  setWord(writer, at + slot, bits);

export const boolBit = (writer: Writer, at: number, slot: number, bit: number, value: boolean) => {
  const idx = at + slot;
  const current = writer.body[idx];
  if (current === undefined || bit < 0 || bit >= WORD_BITS) {
    return tdbinErr<undefined>("LimitExceeded");
  }
  const mask = 1n << BigInt(bit);
  writer.body[idx] = value ? current | mask : current & ~mask;
  return ok(undefined);
};

export const string = (writer: Writer, at: number, dataWords: number, slot: number, value: string | null) =>
  bytes(writer, at, dataWords, slot, value === null ? null : utf8Encode(value));

export const bytes = (
  writer: Writer,
  at: number,
  dataWords: number,
  slot: number,
  value: Uint8Array | null
): Result<void, TdbinError> => {
  const ptrWord = ptrIndex(at, dataWords, slot);
  return value === null ? setWord(writer, ptrWord, 0n) : writeByteList(writer, ptrWord, value);
};

export const boolList = (
  writer: Writer,
  at: number,
  dataWords: number,
  slot: number,
  values: readonly boolean[] | null
) => {
  const ptrWord = ptrIndex(at, dataWords, slot);
  return values === null ? setWord(writer, ptrWord, 0n) : writeBoolList(writer, ptrWord, values);
};

export const byteList = bytes;

export const wordList = (
  writer: Writer,
  at: number,
  dataWords: number,
  slot: number,
  values: readonly bigint[] | null
): Result<void, TdbinError> => {
  const ptrWord = ptrIndex(at, dataWords, slot);
  return values === null ? setWord(writer, ptrWord, 0n) : writeWordList(writer, ptrWord, values);
};

export const bytes16List = (
  writer: Writer,
  at: number,
  dataWords: number,
  slot: number,
  values: readonly (readonly [bigint, bigint])[] | null
): Result<void, TdbinError> => {
  const ptrWord = ptrIndex(at, dataWords, slot);
  return values === null ? setWord(writer, ptrWord, 0n) : writeBytes16List(writer, ptrWord, values);
};

export const stringList = (
  writer: Writer,
  at: number,
  dataWords: number,
  slot: number,
  values: readonly string[] | null
) => {
  const ptrWord = ptrIndex(at, dataWords, slot);
  return values === null ? setWord(writer, ptrWord, 0n) : writePointerList(writer, ptrWord, values, writeStringPointer);
};

export const bytesList = (
  writer: Writer,
  at: number,
  dataWords: number,
  slot: number,
  values: readonly Uint8Array[] | null
): Result<void, TdbinError> => {
  const ptrWord = ptrIndex(at, dataWords, slot);
  return values === null ? setWord(writer, ptrWord, 0n) : writePointerList(writer, ptrWord, values, writeBytesPointer);
};

export const child = <T>(
  writer: Writer,
  at: number,
  dataWords: number,
  slot: number,
  codec: StructCodec<T>,
  value: T | null
): Result<void, TdbinError> => {
  const ptrWord = ptrIndex(at, dataWords, slot);
  return value === null ? setWord(writer, ptrWord, 0n) : writeChild(writer, ptrWord, codec, value);
};

export const childList = <T>(
  writer: Writer,
  at: number,
  dataWords: number,
  slot: number,
  codec: StructCodec<T>,
  values: readonly T[] | null
): Result<void, TdbinError> => {
  const ptrWord = ptrIndex(at, dataWords, slot);
  return values === null ? setWord(writer, ptrWord, 0n) : writeChildList(writer, ptrWord, codec, values);
};

const reserve = (writer: Writer, words: number): Result<number, TdbinError> => {
  const start = writer.body.length;
  const end = start + words;
  if (!Number.isSafeInteger(end) || end > MAX_WORDS) {
    return tdbinErr("LimitExceeded");
  }
  writer.body.length = end;
  writer.body.fill(0n, start, end);
  return ok(start);
};

const setWord = (writer: Writer, idx: number, value: bigint): Result<void, TdbinError> => {
  if (idx < 0 || idx >= writer.body.length) {
    return tdbinErr("LimitExceeded");
  }
  writer.body[idx] = BigInt.asUintN(64, value);
  return ok(undefined);
};

const ptrIndex = (at: number, dataWords: number, slot: number): number => at + dataWords + slot;

const structPtr = (target: number, ptrWord: number, dataWords: number, ptrWords: number) => {
  const offset = relOffset(target, ptrWord);
  return offset.ok ? encodeStruct(offset.value, dataWords, ptrWords) : offset;
};

const listPtr = (target: number, ptrWord: number, elem: number, count: number) => {
  const offset = relOffset(target, ptrWord);
  return offset.ok ? encodeList(offset.value, elem, count) : offset;
};

const setListPtr = (writer: Writer, ptrWord: number, target: number, elem: number, count: number) => {
  const ptr = listPtr(target, ptrWord, elem, count);
  return ptr.ok ? setWord(writer, ptrWord, ptr.value) : ptr;
};

const writeByteList = (writer: Writer, ptrWord: number, data: Uint8Array) => {
  const words = Math.ceil(data.length / WORD_BYTES);
  const start = reserve(writer, words);
  const packed = start.ok ? packBytes(writer, start.value, data) : start;
  return start.ok && packed.ok ? setListPtr(writer, ptrWord, start.value, ELEM_BYTE, data.length) : packed;
};

const writeBoolList = (writer: Writer, ptrWord: number, values: readonly boolean[]) => {
  const start = reserve(writer, Math.ceil(values.length / WORD_BITS));
  const packed = start.ok ? packBools(writer, start.value, values) : start;
  return start.ok && packed.ok ? setListPtr(writer, ptrWord, start.value, ELEM_BIT, values.length) : packed;
};

const writeWordList = (writer: Writer, ptrWord: number, values: readonly bigint[]) => {
  const start = reserve(writer, values.length);
  const copied = start.ok ? copyWords(writer, start.value, values) : start;
  return start.ok && copied.ok ? setListPtr(writer, ptrWord, start.value, ELEM_EIGHT_BYTES, values.length) : copied;
};

const writePointerList = <T>(
  writer: Writer,
  ptrWord: number,
  values: readonly T[],
  writeOne: (writer: Writer, ptrWord: number, value: T) => Result<void, TdbinError>
) => {
  const start = reserve(writer, values.length);
  const ptr = start.ok ? setListPtr(writer, ptrWord, start.value, ELEM_POINTER, values.length) : start;
  return start.ok && ptr.ok ? writeEachPointer(writer, start.value, values, writeOne) : ptr;
};

const writeChild = <T>(writer: Writer, ptrWord: number, codec: StructCodec<T>, value: T) => {
  const start = reserve(writer, codec.dataWords + codec.ptrWords);
  if (!start.ok) {
    return start;
  }
  const written = codec.write(writer, start.value, value);
  if (!written.ok) {
    return written;
  }
  const ptr = structPtr(start.value, ptrWord, codec.dataWords, codec.ptrWords);
  return ptr.ok ? setWord(writer, ptrWord, ptr.value) : ptr;
};

const writeChildList = <T>(writer: Writer, ptrWord: number, codec: StructCodec<T>, values: readonly T[]) => {
  const stride = codec.dataWords + codec.ptrWords;
  const elemWords = stride * values.length;
  const start =
    stride === 0 && values.length !== 0 ? tdbinErr<number>("LimitExceeded") : reserve(writer, elemWords + 1);
  const tag = start.ok ? writeCompositeTag(writer, start.value, values.length, codec.dataWords, codec.ptrWords) : start;
  const items = start.ok && tag.ok ? writeCompositeItems(writer, start.value, stride, codec, values) : tag;
  return start.ok && items.ok ? setListPtr(writer, ptrWord, start.value, ELEM_COMPOSITE, elemWords) : items;
};

const writeBytes16List = (writer: Writer, ptrWord: number, values: readonly (readonly [bigint, bigint])[]) => {
  const elemWords = values.length * 2;
  const start = reserve(writer, elemWords + 1);
  const tag = start.ok ? writeCompositeTag(writer, start.value, values.length, 2, 0) : start;
  const items = start.ok && tag.ok ? writeBytes16Items(writer, start.value, values) : tag;
  return start.ok && items.ok ? setListPtr(writer, ptrWord, start.value, ELEM_COMPOSITE, elemWords) : items;
};

const writeCompositeTag = (writer: Writer, start: number, count: number, dataWords: number, ptrWords: number) => {
  const tag = encodeStruct(count, dataWords, ptrWords);
  return tag.ok ? setWord(writer, start, tag.value) : tag;
};

const writeCompositeItems = <T>(
  writer: Writer,
  start: number,
  stride: number,
  codec: StructCodec<T>,
  values: readonly T[]
) =>
  values.reduce<Result<void, TdbinError>>(
    (state, value, index) => (state.ok ? codec.write(writer, start + 1 + stride * index, value) : state),
    ok(undefined)
  );

const writeBytes16Items = (writer: Writer, start: number, values: readonly (readonly [bigint, bigint])[]) =>
  values.reduce<Result<void, TdbinError>>((state, value, index) => {
    const first = value[0];
    const second = value[1];
    const at = start + 1 + index * 2;
    const written = state.ok ? setWord(writer, at, first) : state;
    return written.ok ? setWord(writer, at + 1, second) : written;
  }, ok(undefined));

const packBytes = (writer: Writer, start: number, data: Uint8Array) => {
  for (let offset = 0; offset < data.length; offset += WORD_BYTES) {
    const chunk = data.slice(offset, offset + WORD_BYTES);
    const word = new Uint8Array(WORD_BYTES);
    word.set(chunk);
    const view = new DataView(word.buffer);
    writer.body[start + offset / WORD_BYTES] = view.getBigUint64(0, true);
  }
  return ok(undefined);
};

const packBools = (writer: Writer, start: number, values: readonly boolean[]) =>
  values.reduce<Result<void, TdbinError>>((state, value, index) => {
    const word = Math.floor(index / WORD_BITS);
    const bit = index % WORD_BITS;
    return state.ok && value ? boolBit(writer, start, word, bit, true) : state;
  }, ok(undefined));

const copyWords = (writer: Writer, start: number, values: readonly bigint[]) =>
  values.reduce<Result<void, TdbinError>>(
    (state, word, index) => (state.ok ? setWord(writer, start + index, word) : state),
    ok(undefined)
  );

const writeEachPointer = <T>(
  writer: Writer,
  start: number,
  values: readonly T[],
  writeOne: (writer: Writer, ptrWord: number, value: T) => Result<void, TdbinError>
) =>
  values.reduce<Result<void, TdbinError>>(
    (state, value, index) => (state.ok ? writeOne(writer, start + index, value) : state),
    ok(undefined)
  );

const writeStringPointer = (writer: Writer, ptrWord: number, value: string) =>
  writeByteList(writer, ptrWord, utf8Encode(value));

const writeBytesPointer = (writer: Writer, ptrWord: number, value: Uint8Array) => writeByteList(writer, ptrWord, value);

File: crates/tdbin/src/writer.rs
//! The word-arena encoder: preorder allocation with in-message back-patching
//! ([TDBIN-ENC-ORDER]); identical values produce byte-identical bodies
//! ([TDBIN-ENC-CANON]). Generated ADT code calls the public methods; the
//! private helpers keep the pointer math in one place. List encoding lives in
//! `writer_lists.rs`; columnar encoding in `column.rs`.
//!
//! The arena is a flat little-endian byte vector, so finishing a message is
//! free and string/list bodies are single bulk copies. An optional byte
//! `prefix` reserves room for a frame header so framed encodes never re-copy
//! the body.

use crate::error::EncodeError;
use crate::layout::WORD_BYTES;
use crate::pointer::{self, ELEM_BYTE};
use crate::{Struct, MAX_DEPTH};

/// Upper bound on message body words (a safety cap for the encoder).
const MAX_WORDS: usize = 1 << 26;
/// Initial arena capacity: covers small messages with one allocation.
const INITIAL_CAPACITY: usize = 256;

/// Accumulates message body words while encoding a value tree.
#[derive(Debug)]
pub struct Writer {
    /// The message: `prefix` reserved bytes then the little-endian body.
    body: Vec<u8>,
    /// Byte offset where body word 0 starts.
    prefix: usize,
    /// Remaining recursive child-pointer depth.
    depth: u32,
}

impl Writer {
    /// Create an empty writer with `prefix` reserved header bytes.
    fn new(prefix: usize) -> Self {
        let mut body = Vec::with_capacity(INITIAL_CAPACITY.max(prefix));
        body.resize(prefix, 0);
        Self {
            body,
            prefix,
            depth: MAX_DEPTH,
        }
    }

    /// Encode a root value into a complete message ([TDBIN-MSG-BARE]).
    ///
    /// # Errors
    /// Returns [`EncodeError`] if the value exceeds a wire-format limit.
    pub(crate) fn message<T: Struct>(value: &T) -> Result<Vec<u8>, EncodeError> {
        Self::message_with_prefix(value, 0)
    }

    /// Encode a message after `prefix` reserved zero bytes ([TDBIN-MSG-FRAME]).
    pub(crate) fn message_with_prefix<T: Struct>(
        value: &T,
        prefix: usize,
    ) -> Result<Vec<u8>, EncodeError> {
        let mut writer = Self::new(prefix);
        let _root = writer.reserve(1)?;
        let words = T::body_words().ok_or(EncodeError::LimitExceeded)?;
        let root_at = match words {
            0 => 0,
            _ => writer.reserve(words)?,
        };
        value.write_struct(&mut writer, root_at)?;
        let offset = rel_offset(root_at, 0)?;
        let ptr = pointer::encode_struct(offset, T::DATA_WORDS, T::PTR_WORDS)?;
        writer.set(0, ptr)?;
        Ok(writer.body)
    }

    /// Reserve `words` zeroed words, returning the start word index.
    pub(crate) fn reserve(&mut self, words: usize) -> Result<usize, EncodeError> {
        let start = self.next_word()?;
        let end = self.end_of(start, words)?;
        self.grow_for(end);
        self.body.resize(end, 0);
        Ok(start)
    }

    /// Append `data` as one whole reservation of `words` words: a single
    /// bulk copy plus a zeroed padding tail, never touching memory twice.
    pub(crate) fn append_reserved(
        &mut self,
        words: usize,
        data: &[u8],
    ) -> Result<usize, EncodeError> {
        let start = self.next_word()?;
        let end = self.end_of(start, words)?;
        let data_end = self
            .body
            .len()
            .checked_add(data.len())
            .filter(|byte| *byte <= end)
            .ok_or(EncodeError::LimitExceeded)?;
        self.grow_for(end);
        self.body.extend_from_slice(data);
        self.body.resize(end.max(data_end), 0);
        Ok(start)
    }

    /// Append every row's bytes end-to-end as one reservation: pure appends,
    /// no pre-zeroing, one zeroed padding tail.
    pub(crate) fn append_concat<'v>(
        &mut self,
        total: usize,
        values: impl Iterator<Item = &'v [u8]>,
    ) -> Result<usize, EncodeError> {
        let start = self.next_word()?;
        let end = self.end_of(start, total.div_ceil(WORD_BYTES))?;
        self.grow_for(end);
        let expected = self
            .body
            .len()
            .checked_add(total)
            .ok_or(EncodeError::LimitExceeded)?;
        for row in values {
            self.body.extend_from_slice(row);
        }
        if self.body.len() != expected {
            return Err(EncodeError::LimitExceeded);
        }
        self.body.resize(end, 0);
        Ok(start)
    }

    /// The next unreserved word index.
    fn next_word(&self) -> Result<usize, EncodeError> {
        Ok(
            (self.body.len().checked_sub(self.prefix)).ok_or(EncodeError::LimitExceeded)?
                / WORD_BYTES,
        )
    }

    /// The byte length after reserving `words` words at `start`.
    fn end_of(&self, start: usize, words: usize) -> Result<usize, EncodeError> {
        let end_words = start.checked_add(words).ok_or(EncodeError::LimitExceeded)?;
        if end_words > MAX_WORDS {
            return Err(EncodeError::LimitExceeded);
        }
        end_words
            .checked_mul(WORD_BYTES)
            .and_then(|bytes| bytes.checked_add(self.prefix))
            .ok_or(EncodeError::LimitExceeded)
    }

    /// Grow capacity ahead of `end` aggressively so bulk encodes do not pay
    /// repeated doubling copies.
    fn grow_for(&mut self, end: usize) {
        if end > self.body.capacity() {
            let ahead = end.max(self.body.capacity().wrapping_mul(4));
            self.body.reserve(ahead.saturating_sub(self.body.len()));
        }
    }

    /// Mutable view of the word at absolute index `idx`.
    fn word_mut(&mut self, idx: usize) -> Result<&mut [u8], EncodeError> {
        self.bytes_mut(idx, WORD_BYTES)
    }

    /// Mutable view of `len` body bytes starting at word `idx`.
    pub(crate) fn bytes_mut(&mut self, idx: usize, len: usize) -> Result<&mut [u8], EncodeError> {
        let start = idx
            .checked_mul(WORD_BYTES)
            .and_then(|bytes| bytes.checked_add(self.prefix))
            .ok_or(EncodeError::LimitExceeded)?;
        let end = start.checked_add(len).ok_or(EncodeError::LimitExceeded)?;
        self.body
            .get_mut(start..end)
            .ok_or(EncodeError::LimitExceeded)
    }

    /// Overwrite the word at absolute index `idx`.
    pub(crate) fn set(&mut self, idx: usize, value: u64) -> Result<(), EncodeError> {
        self.word_mut(idx)?.copy_from_slice(&value.to_le_bytes());
        Ok(())
    }

    /// Write a scalar into data `slot` of the struct at `at` ([TDBIN-REC-ALLOC]).
    ///
    /// # Errors
    /// Returns [`EncodeError`] if the slot index overflows.
    pub fn scalar(&mut self, at: usize, slot: u16, bits: u64) -> Result<(), EncodeError> {
        let idx = at
            .checked_add(usize::from(slot))
            .ok_or(EncodeError::LimitExceeded)?;
        self.set(idx, bits)
    }

    /// Write one packed Bool bit into a data word ([TDBIN-REC-XOR]).
    ///
    /// # Errors
    /// Returns [`EncodeError`] if the slot index overflows or the bit is invalid.
    pub fn bool_bit(
        &mut self,
        at: usize,
        slot: u16,
        bit: u8,
        value: bool,
    ) -> Result<(), EncodeError> {
        let idx = at
            .checked_add(usize::from(slot))
            .ok_or(EncodeError::LimitExceeded)?;
        let mask = 1_u64
            .checked_shl(u32::from(bit))
            .ok_or(EncodeError::LimitExceeded)?;
        let cell = self.word_mut(idx)?;
        let word = word_of(cell)?;
        let updated = if value { word | mask } else { word & !mask };
        cell.copy_from_slice(&updated.to_le_bytes());
        Ok(())
    }

    /// Write an optional UTF-8 string into pointer `slot` ([TDBIN-PRIM-MAP]).
    ///
    /// # Errors
    /// Returns [`EncodeError`] if the message exceeds a wire-format limit.
    pub fn string(
        &mut self,
        at: usize,
        data_words: u16,
        slot: u16,
        value: Option<&str>,
    ) -> Result<(), EncodeError> {
        let ptr_word = Self::ptr_index(at, data_words, slot)?;
        match value {
            None => self.set(ptr_word, 0),
            Some(text) => self.write_byte_list(ptr_word, text.as_bytes()),
        }
    }

    /// Write optional raw bytes into pointer `slot` ([TDBIN-PRIM-MAP]).
    ///
    /// # Errors
    /// Returns [`EncodeError`] if the message exceeds a wire-format limit.
    pub fn bytes(
        &mut self,
        at: usize,
        data_words: u16,
        slot: u16,
        value: Option<&[u8]>,
    ) -> Result<(), EncodeError> {
        self.byte_list(at, data_words, slot, value)
    }

    /// Write an optional child struct into pointer `slot` ([TDBIN-PTR-STRUCT]).
    ///
    /// # Errors
    /// Returns [`EncodeError`] if the message exceeds a wire-format limit.
    pub fn child<C: Struct>(
        &mut self,
        at: usize,
        data_words: u16,
        slot: u16,
        value: Option<&C>,
    ) -> Result<(), EncodeError> {
        let ptr_word = Self::ptr_index(at, data_words, slot)?;
        match value {
            None => self.set(ptr_word, 0),
            Some(child) => self.write_child(ptr_word, child),
        }
    }

    /// Absolute word index of pointer `slot`.
    pub(crate) fn ptr_index(at: usize, data_words: u16, slot: u16) -> Result<usize, EncodeError> {
        crate::layout::ptr_word(at, data_words, usize::from(slot)).ok_or(EncodeError::LimitExceeded)
    }

    /// Append a child struct body and patch its pointer word.
    fn write_child<C: Struct>(&mut self, ptr_word: usize, child: &C) -> Result<(), EncodeError> {
        let words = C::body_words().ok_or(EncodeError::LimitExceeded)?;
        let child_at = match words {
            0 => ptr_word,
            _ => self.reserve(words)?,
        };
        self.with_descended(|writer| child.write_struct(writer, child_at))?;
        let offset = rel_offset(child_at, ptr_word)?;
        let ptr = pointer::encode_struct(offset, C::DATA_WORDS, C::PTR_WORDS)?;
        self.set(ptr_word, ptr)
    }

    /// Append a byte-list body with one bulk copy and patch its pointer word.
    pub(crate) fn write_byte_list(
        &mut self,
        ptr_word: usize,
        data: &[u8],
    ) -> Result<(), EncodeError> {
        let words = data.len().div_ceil(WORD_BYTES);
        let start = self.append_reserved(words, data)?;
        self.set_list_ptr(ptr_word, start, ELEM_BYTE, data.len())
    }

    /// Patch a list pointer word after appending its body.
    pub(crate) fn set_list_ptr(
        &mut self,
        ptr_word: usize,
        start: usize,
        elem: u8,
        count: usize,
    ) -> Result<(), EncodeError> {
        let offset = rel_offset(start, ptr_word)?;
        let ptr = pointer::encode_list(offset, elem, count)?;
        self.set(ptr_word, ptr)
    }

    /// Pack bits little-endian into already-reserved words at `start`.
    pub(crate) fn pack_bits(
        &mut self,
        start: usize,
        values: impl Iterator<Item = bool>,
    ) -> Result<(), EncodeError> {
        let byte_start = start
            .checked_mul(WORD_BYTES)
            .and_then(|bytes| bytes.checked_add(self.prefix))
            .ok_or(EncodeError::LimitExceeded)?;
        let dst = self
            .body
            .get_mut(byte_start..)
            .ok_or(EncodeError::LimitExceeded)?;
        for (i, value) in values.enumerate() {
            let cell = dst.get_mut(i / 8).ok_or(EncodeError::LimitExceeded)?;
            let mask = 1_u8.wrapping_shl(u32::try_from(i % 8).unwrap_or(0));
            *cell |= mask & u8::from(value).wrapping_neg();
        }
        Ok(())
    }

    /// Run one nested struct write with the pointer-depth budget decremented.
    pub(crate) fn with_descended<T>(
        &mut self,
        write: impl FnOnce(&mut Self) -> Result<T, EncodeError>,
    ) -> Result<T, EncodeError> {
        let previous = self.depth;
        self.depth = previous.checked_sub(1).ok_or(EncodeError::LimitExceeded)?;
        let result = write(self);
        self.depth = previous;
        result
    }
}

/// Read a word back out of a mutable cell.
fn word_of(cell: &[u8]) -> Result<u64, EncodeError> {
    <[u8; WORD_BYTES]>::try_from(cell)
        .map(u64::from_le_bytes)
        .map_err(|_| EncodeError::LimitExceeded)
}

/// Relative offset (in words) from the end of a pointer word to a target.
pub(crate) fn rel_offset(target_word: usize, ptr_word: usize) -> Result<i64, EncodeError> {
    let target = i64::try_from(target_word).map_err(|_| EncodeError::LimitExceeded)?;
    let base = i64::try_from(ptr_word)
        .map_err(|_| EncodeError::LimitExceeded)?
        .checked_add(1)
        .ok_or(EncodeError::LimitExceeded)?;
    target.checked_sub(base).ok_or(EncodeError::LimitExceeded)
}
